File: src/bmsx/action_effects/input_action_effect_compiler.ts
import { has_own } from '../utils/has_own';
import type { PlayerInput } from '../input/playerinput';
import type {
	InputActionEffectProgram,
	Binding,
	Effect,
	ActionEffectTriggerDescriptor,
	EmitGameplayDescriptor,
} from './input_action_effect_dsl';
import { create_gameevent, type GameEvent } from '../core/game_event';
import type { ActionEffectId } from './effect_types';
import type { ActionEffectComponent } from '../component/actioneffectcomponent';
import type { WorldObject } from '../core/object/worldobject';
import { ActionEffectRegistry } from './effect_registry';

export interface BindingExecutionEnv {
	owner: WorldObject;
	ownerId: string;
	playerIndex: number;
	input: PlayerInput;
	effects?: ActionEffectComponent;
	queuedEvents: GameEvent[];
}

export type PatternPredicate = (input: PlayerInput) => boolean;
export type EffectExecutor = (env: BindingExecutionEnv) => void;

export interface CompiledCustomEdge {
	name: string;
	match: PatternPredicate;
	effect?: EffectExecutor;
}

type BindingAnalysis = {
	usesEffectTriggers: boolean;
};

export interface CompiledBinding {
	name?: string;
	priority: number;
	predicate: (env: BindingExecutionEnv) => boolean;
	press?: PatternPredicate;
	hold?: PatternPredicate;
	release?: PatternPredicate;
	pressEffect?: EffectExecutor;
	holdEffect?: EffectExecutor;
	releaseEffect?: EffectExecutor;
	customEdges: CompiledCustomEdge[];
	usesEffectTriggers: boolean;
}

export interface CompiledProgram {
	evalMode: 'first' | 'all';
	priority: number;
	bindings: CompiledBinding[];
	usesEffectTriggers: boolean;
}

export type PatternParser = (pattern: string) => PatternPredicate;

export function compileProgram(program: InputActionEffectProgram, parse: PatternParser): CompiledProgram {
	const progPriority = program.priority ?? 0;
	const evalMode = program.eval ?? 'first';
	const bindings = program.bindings ?? [];

	const compiledEntries = bindings.map((binding, index) => ({
		index,
		compiled: compileBinding(binding, parse),
	}));

	compiledEntries.sort((a, b) => {
		const prio = b.compiled.priority - a.compiled.priority;
		if (prio !== 0) return prio;
		return a.index - b.index;
	});

	let usesEffectTriggers = false;
	for (let i = 0; i < compiledEntries.length; i++) {
		const entry = compiledEntries[i]!;
		if (entry.compiled.usesEffectTriggers) usesEffectTriggers = true;
	}

	return {
		evalMode,
		priority: progPriority,
		bindings: compiledEntries.map(entry => entry.compiled),
		usesEffectTriggers,
	};
}

function compileBinding(binding: Binding, parse: PatternParser): CompiledBinding {
	const priority = binding.priority ?? 0;
	const analysis: BindingAnalysis = {
		usesEffectTriggers: false,
	};
	const predicate = compilePredicate(binding);
	if (!binding.on) {
		throw new Error(`[InputActionEffectCompiler] Binding '${binding.name ?? '(unnamed)'}' is missing an 'on' clause.`);
	}
	const press = binding.on.press ? parse(binding.on.press) : undefined;
	const hold = binding.on.hold ? parse(binding.on.hold) : undefined;
	const release = binding.on.release ? parse(binding.on.release) : undefined;
	const customEntries = binding.on.custom ?? [];
	const customEffects = compileCustomEffects(binding, analysis);
	const customEdges = customEntries.map(item => ({
		name: item.name,
		match: parse(item.pattern),
		effect: customEffects.get(item.name),
	}));

	return {
		name: binding.name,
		priority,
		predicate,
		press,
		hold,
		release,
		pressEffect: compileEffectList(binding.do?.press, 'press', analysis),
		holdEffect: compileEffectList(binding.do?.hold, 'hold', analysis),
		releaseEffect: compileEffectList(binding.do?.release, 'release', analysis),
		customEdges,
		usesEffectTriggers: analysis.usesEffectTriggers,
	};
}

function compilePredicate(binding: Binding): (env: BindingExecutionEnv) => boolean {
	const when = binding.when;
	if (!when) return () => true;

	const modePred = when.mode;
	const modeItems = modePred ? (Array.isArray(modePred) ? modePred : [modePred]) : undefined;
	if (modeItems) {
		for (let i = 0; i < modeItems.length; i++) {
			const item = modeItems[i]!;
			if (!item.path) {
				throw new Error(`[InputActionEffectCompiler] 'mode' clause missing 'path' in binding '${binding.name ?? '(unnamed)'}'.`);
			}
		}
	}

	if (!modeItems) return () => true;

	return (env: BindingExecutionEnv) => {
		for (let i = 0; i < modeItems.length; i++) {
			const entry = modeItems[i]!;
			const entryPath = entry.path!;
			const matches = env.owner.sc.matches_state_path(entryPath);
			if (entry.not) {
				if (matches) return false;
			} else if (!matches) {
				return false;
			}
		}
		return true;
	};
}

function compileCustomEffects(binding: Binding, analysis: BindingAnalysis): Map<string, EffectExecutor> {
	const map = new Map<string, EffectExecutor>();
	const table = binding.do ?? {};
	for (const key of Object.keys(table)) {
		if (key === 'press' || key === 'hold' || key === 'release') continue;
		map.set(key, compileEffectList(table[key], key, analysis));
	}
	return map;
}

function compileEffectList(spec: Effect | Effect[], slot?: string, analysis?: BindingAnalysis): EffectExecutor {
	if (!spec) return undefined;
	const entries = Array.isArray(spec) ? spec : [spec];
	const executors: EffectExecutor[] = [];
	for (let i = 0; i < entries.length; i++) {
		executors.push(compileEffect(entries[i]!, slot, analysis));
	}
	if (executors.length === 0) throw new Error(`Empty effect list in slot '${slot ?? 'unknown'}'.`);
	if (executors.length === 1) return executors[0];
	return (env: BindingExecutionEnv) => {
		for (let i = 0; i < executors.length; i++) {
			executors[i](env);
		}
	};
}

function compileEffect(effect: Effect, slot?: string, analysis?: BindingAnalysis): EffectExecutor {
	if (isEffectTrigger(effect)) {
		if (analysis) analysis.usesEffectTriggers = true;
		const spec = effect['effect.trigger'];
		if (!spec) throw new Error(`Missing effect trigger in effect ${JSON.stringify(effect)}`);
		if (typeof spec === 'string') {
			return (env: BindingExecutionEnv) => {
				executeEffectTrigger(env, spec as ActionEffectId);
			};
		}
		return (env: BindingExecutionEnv) => {
			if (spec.payload === undefined) executeEffectTrigger(env, spec.id as ActionEffectId);
			else executeEffectTrigger(env, spec.id as ActionEffectId, spec.payload);
		};
	}
	if (isInputConsume(effect)) {
		const actions = Array.isArray(effect['input.consume']) ? effect['input.consume'] : [effect['input.consume']];
		if (actions.length === 0) throw new Error(`Empty actions in input.consume effect ${JSON.stringify(effect)}`);
		return (env: BindingExecutionEnv) => {
			for (let i = 0; i < actions.length; i++) {
				env.input.consumeAction(actions[i]!);
			}
		};
	}
	if (isGameplayEmit(effect)) {
		const { event, payload } = effect['emit.gameplay'];
		if (!event) throw new Error(`Missing event name in emit.gameplay effect ${JSON.stringify(effect)}`);
		return (env: BindingExecutionEnv) => {
			const evt = create_gameevent({ type: event, ...(payload ?? {}) });
			env.queuedEvents.push(evt);
		};
	}
	if (isNestedCommands(effect)) {
		const nested = compileEffectList(effect.commands, slot, analysis);
		if (!nested) throw new Error(`Empty commands in nested effect ${JSON.stringify(effect)}`);
		return nested;
	}
	throw new Error(`[InputActionEffectCompiler] Unknown effect in slot '${slot ?? 'unknown'}': ${JSON.stringify(effect)}`);
}

function executeEffectTrigger(env: BindingExecutionEnv, id: ActionEffectId, payload?: unknown) {
	const effects = env.effects;
	if (!effects) {
		throw new Error(`[InputActionEffectCompiler] Effect trigger '${id}' attempted without ActionEffectComponent on '${env.ownerId}'.`);
	}
	const result = effects.trigger(id, payload);
	return result;
}

type ValidationContext = {
	programId: string;
	bindingName: string;
	slot: string;
};

export function validateProgramEffects(program: InputActionEffectProgram, programId: string): void {
	const bindings = program.bindings;
	for (let index = 0; index < bindings.length; index++) {
		const binding = bindings[index]!;
		const bindingName = binding.name ? binding.name : `#${index}`;
		const table = binding.do;
		if (!table) {
			throw new Error(`[InputActionEffectProgramValidation] Program '${programId}' binding '${bindingName}' missing effect table.`);
		}
		validateEffectSpec(table.press, { programId, bindingName, slot: 'press' });
		validateEffectSpec(table.hold, { programId, bindingName, slot: 'hold' });
		validateEffectSpec(table.release, { programId, bindingName, slot: 'release' });
		const keys = Object.keys(table);
		for (let keyIndex = 0; keyIndex < keys.length; keyIndex++) {
			const key = keys[keyIndex]!;
			if (key === 'press' || key === 'hold' || key === 'release') continue;
			const customSpec = table[key];
			if (!customSpec) continue;
			const slot = `custom:${key}`;
			validateEffectSpec(customSpec, { programId, bindingName, slot });
		}
	}
}
function validateEffectSpec(spec: Effect | Effect[], ctx: ValidationContext): void {
	if (Array.isArray(spec)) {
		for (let i = 0; i < spec.length; i++) {
			const entry = spec[i]!;
			const slot = `${ctx.slot}[${i}]`;
			validateEffect(entry, { programId: ctx.programId, bindingName: ctx.bindingName, slot });
		}
		return;
	}
	validateEffect(spec, ctx);
}

function validateEffect(effect: Effect, ctx: ValidationContext): void {
	if (isEffectTrigger(effect)) {
		const descriptor = effect['effect.trigger'];
		if (!descriptor) {
			throw new Error(`[InputActionEffectProgramValidation] Program '${ctx.programId}' binding '${ctx.bindingName}' slot '${ctx.slot}' missing effect trigger descriptor.`);
		}
		let effectId: ActionEffectId;
		let payload: unknown;
		if (typeof descriptor === 'string') {
			effectId = descriptor as ActionEffectId;
			payload = undefined;
		} else {
			const id = descriptor.id;
			if (!id) {
				throw new Error(`[InputActionEffectProgramValidation] Program '${ctx.programId}' binding '${ctx.bindingName}' slot '${ctx.slot}' effect trigger missing id.`);
			}
			effectId = id as ActionEffectId;
			payload = descriptor.payload;
		}
		try {
			ActionEffectRegistry.instance.validate(effectId, payload);
		} catch (error) {
			const message = error instanceof Error ? error.message : String(error);
			throw new Error(`[InputActionEffectProgramValidation] Program '${ctx.programId}' binding '${ctx.bindingName}' slot '${ctx.slot}' effect '${effectId}' validation failed: ${message}`);
		}
		return;
	}
	if (isNestedCommands(effect)) {
		const commands = effect.commands;
		for (let i = 0; i < commands.length; i++) {
			const nested = commands[i]!;
			const slot = `${ctx.slot}.commands[${i}]`;
			validateEffect(nested, { programId: ctx.programId, bindingName: ctx.bindingName, slot });
		}
	}
}

function isEffectTrigger(effect: Effect): effect is { 'effect.trigger': ActionEffectId | ActionEffectTriggerDescriptor } {
	return has_own(effect, 'effect.trigger');
}

function isInputConsume(effect: Effect): effect is { 'input.consume': string | string[] } {
	return has_own(effect, 'input.consume');
}

function isGameplayEmit(effect: Effect): effect is { 'emit.gameplay': EmitGameplayDescriptor } {
	return has_own(effect, 'emit.gameplay');
}

function isNestedCommands(effect: Effect): effect is { commands: Effect[] } {
	return has_own(effect, 'commands');
}

File: src/bmsx/res/systemrom/input_action_effect_compiler.lua
-- input_action_effect_compiler.lua
-- compile input action effect programs into executable bindings

local action_effects = require("action_effects")
local eventemitter = require("eventemitter")

local compile_effect_list

local function is_effect_trigger(effect)
	return effect["effect.trigger"] ~= nil
end

local function is_input_consume(effect)
	return effect["input.consume"] ~= nil
end

local function is_gameplay_emit(effect)
	return effect["emit.gameplay"] ~= nil
end

local function is_nested_commands(effect)
	return effect.commands ~= nil
end

local function execute_effect_trigger(env, id, payload)
	local effects = env.effects
	if not effects then
		error("[inputactioneffectcompiler] effect trigger '" .. id .. "' attempted without actioneffectcomponent on '" .. env.owner_id .. "'.")
	end
	if payload == nil then
		return effects:trigger(id)
	end
	return effects:trigger(id, { payload = payload })
end

local function compile_effect(effect, slot, analysis)
	if is_effect_trigger(effect) then
		if analysis then
			analysis.uses_effect_triggers = true
		end
		local spec = effect["effect.trigger"]
		if type(spec) == "string" then
			return function(env)
				execute_effect_trigger(env, spec)
			end
		end
		return function(env)
			execute_effect_trigger(env, spec.id, spec.payload)
		end
	end
	if is_input_consume(effect) then
		local actions = effect["input.consume"]
		if type(actions) ~= "table" then
			actions = { actions }
		end
		return function(env)
			for i = 1, #actions do
				$.consume_action(env.player_index, actions[i])
			end
		end
	end
	if is_gameplay_emit(effect) then
		local spec = effect["emit.gameplay"]
		return function(env)
			local payload = spec.payload or {}
			local base = { type = spec.event }
			for k, v in pairs(payload) do
				base[k] = v
			end
			env.queued_events[#env.queued_events + 1] = eventemitter.create_gameevent(base)
		end
	end
	if is_nested_commands(effect) then
		local nested = compile_effect_list(effect.commands, slot, analysis)
		return nested
	end
	error("[inputactioneffectcompiler] unknown effect in slot '" .. (slot or "unknown") .. "'.")
end

local function compile_effect_list(spec, slot, analysis)
	if not spec then
		return nil
	end
	local entries
	if type(spec) == "table" and spec[1] ~= nil then
		entries = spec
	else
		entries = { spec }
	end
	local executors = {}
	for i = 1, #entries do
		executors[#executors + 1] = compile_effect(entries[i], slot, analysis)
	end
	if #executors == 1 then
		return executors[1]
	end
	return function(env)
		for i = 1, #executors do
			executors[i](env)
		end
	end
end

local function compile_predicate(binding)
	local when = binding.when
	if not when then
		return function()
			return true
		end
	end
	local mode_pred = when.mode
	local mode_items = nil
	if mode_pred then
		if type(mode_pred) == "table" then
			mode_items = mode_pred
		else
			mode_items = { mode_pred }
		end
	end
	if not mode_items then
		return function()
			return true
		end
	end
	return function(env)
		for i = 1, #mode_items do
			local entry = mode_items[i]
			local matches = env.owner.sc:matches_state_path(entry.path)
			if entry["not"] then
				if matches then
					return false
				end
			else
				if not matches then
					return false
				end
			end
		end
		return true
	end
end

local function compile_custom_effects(binding, analysis)
	local map = {}
	local table_go = binding.go or {}
	for key, spec in pairs(table_go) do
		if key ~= "press" and key ~= "hold" and key ~= "release" then
			map[key] = compile_effect_list(spec, key, analysis)
		end
	end
	return map
end

local function compile_binding(binding, parse)
	local priority = binding.priority or 0
	local analysis = { uses_effect_triggers = false }
	local predicate = compile_predicate(binding)
	local on = binding.on
	if not on then
		error("[inputactioneffectcompiler] binding '" .. (binding.name or "(unnamed)") .. "' is missing an 'on' clause.")
	end
	local press = on.press and parse(on.press) or nil
	local hold = on.hold and parse(on.hold) or nil
	local release = on.release and parse(on.release) or nil
	local custom_entries = on.custom or {}
	local custom_effects = compile_custom_effects(binding, analysis)
	local custom_edges = {}
	for i = 1, #custom_entries do
		local entry = custom_entries[i]
		custom_edges[#custom_edges + 1] = {
			name = entry.name,
			match = parse(entry.pattern),
			effect = custom_effects[entry.name],
		}
	end

	return {
		name = binding.name,
		priority = priority,
		predicate = predicate,
		press = press,
		hold = hold,
		release = release,
		press_effect = compile_effect_list(binding.go and binding.go.press or nil, "press", analysis),
		hold_effect = compile_effect_list(binding.go and binding.go.hold or nil, "hold", analysis),
		release_effect = compile_effect_list(binding.go and binding.go.release or nil, "release", analysis),
		custom_edges = custom_edges,
		uses_effect_triggers = analysis.uses_effect_triggers,
	}
end

function compile_program(program, parse)
	local prog_priority = program.priority or 0
	local eval_mode = program.eval or "first"
	local bindings = program.bindings or {}

	local compiled_entries = {}
	for i = 1, #bindings do
		compiled_entries[#compiled_entries + 1] = {
			index = i,
			compiled = compile_binding(bindings[i], parse),
		}
	end

	table.sort(compiled_entries, function(a, b)
		if a.compiled.priority ~= b.compiled.priority then
			return a.compiled.priority > b.compiled.priority
		end
		return a.index < b.index
	end)

	local uses_effect_triggers = false
	for i = 1, #compiled_entries do
		if compiled_entries[i].compiled.uses_effect_triggers then
			uses_effect_triggers = true
			break
		end
	end

	local out_bindings = {}
	for i = 1, #compiled_entries do
		out_bindings[#out_bindings + 1] = compiled_entries[i].compiled
	end

	return {
		eval_mode = eval_mode,
		priority = prog_priority,
		bindings = out_bindings,
		uses_effect_triggers = uses_effect_triggers,
	}
end

local function validate_effect(effect, ctx)
	if not effect then
		return
	end
	if is_effect_trigger(effect) then
		local descriptor = effect["effect.trigger"]
		local effect_id = nil
		local payload = nil
		if type(descriptor) == "string" then
			effect_id = descriptor
		else
			effect_id = descriptor.id
			payload = descriptor.payload
		end
		action_effects.validate(effect_id, payload)
		return
	end
	if is_nested_commands(effect) then
		local commands = effect.commands
		for i = 1, #commands do
			local slot = ctx.slot .. ".commands[" .. i .. "]"
			validate_effect(commands[i], { program_id = ctx.program_id, binding_name = ctx.binding_name, slot = slot })
		end
	end
end

local function validate_effect_spec(spec, ctx)
	if not spec then
		return
	end
	if type(spec) == "table" and spec[1] ~= nil then
		for i = 1, #spec do
			local slot = ctx.slot .. "[" .. i .. "]"
			validate_effect(spec[i], { program_id = ctx.program_id, binding_name = ctx.binding_name, slot = slot })
		end
		return
	end
	validate_effect(spec, ctx)
end

function validate_program_effects(program, program_id)
	local bindings = program.bindings or {}
	for i = 1, #bindings do
		local binding = bindings[i]
		local binding_name = binding.name or ("#" .. i)
		local table_go = binding.go
		if not table_go then
			error("[inputactioneffectprogramvalidation] program '" .. program_id .. "' binding '" .. binding_name .. "' missing effect table.")
		end
		validate_effect_spec(table_go.press, { program_id = program_id, binding_name = binding_name, slot = "press" })
		validate_effect_spec(table_go.hold, { program_id = program_id, binding_name = binding_name, slot = "hold" })
		validate_effect_spec(table_go.release, { program_id = program_id, binding_name = binding_name, slot = "release" })
		for key, spec in pairs(table_go) do
			if key ~= "press" and key ~= "hold" and key ~= "release" then
				validate_effect_spec(spec, { program_id = program_id, binding_name = binding_name, slot = "custom:" .. key })
			end
		end
	end
end

return {
	compile_program = compile_program,
	compile_effect_list = compile_effect_list,
	validate_program_effects = validate_program_effects,
}
